File: src/main/java/lab3/entity/Create.java
package lab3.entity;

import java.util.List;
import java.util.function.Function;
import java.util.function.Supplier;

public class Create extends DelayedTask {

    private final Function<List<Process>, Process> router;
    private final List<Process> consumers;
    private final Supplier<Integer> typeProducer;

    private int createdEvents;

    public Create(
            double initialDelay,
            Supplier<Double> delayGenerator,
            List<Process> consumers,
            Function<List<Process>, Process> router,
            Supplier<Integer> typeProducer
    ) {
        super(delayGenerator);
        this.nextEventTime = initialDelay;
        this.router = router;
        this.consumers = consumers;
        this.typeProducer = typeProducer;
    }

    public Create(
            double initialDelay,
            Supplier<Double> delayGenerator,
            List<Process> consumers,
            Function<List<Process>, Process> router
    ) {
        this(initialDelay, delayGenerator, consumers, router, () -> 1);
    }

    public void processEvent() {
        Entity entity = new Entity();
        entity.setCreationTime(nextEventTime);
        entity.setType(typeProducer.get());
        router.apply(consumers).accept(entity, nextEventTime);

        createdEvents++;
        nextEventTime += delayGenerator.get();
    }

    @Override
    public void calculateStats(double delta) {}

    public void printInfo() {
        System.out.println("Created " + createdEvents + " events");
    }
}

File: src/main/java/lab3/entity/kotlin/Create.kt
package lab3.entity.kotlin

import lab3.entity.DelayedTask
import lab3.entity.Entity

class Create(
    delayGenerator: (Int) -> Double,
    private val consumers: List<Operator>,
    private val router: (List<Operator>) -> Operator = { it.first() },
    private val typeProducer: () -> Int = {1},
    private val maxElementsNum: Int = Int.MAX_VALUE,
    initialDelay: Double = 0.0
): DelayedTask(delayGenerator) {

    private var nextEventTime = initialDelay
    private var createdEvents: Int = 0

    override fun processEvent(currentTime: Double) {
        val type = typeProducer()
        val entity = Entity(type, nextEventTime)
        router(consumers).accept(entity, nextEventTime)

        createdEvents++
        if (createdEvents < maxElementsNum)
            nextEventTime += delayGenerator.apply(0)
        else nextEventTime = Double.MAX_VALUE
    }

    override fun getFinishEventProcessingTime() = nextEventTime

    override fun calculateStats(delta: Double) {}

    fun printInfo() = println("Created $createdEvents events")
}
